README.md
A simple & lightweight method of displaying modal windows with jQuery.

For quick examples and demos, head to [jquerymodal.com](http://jquerymodal.com/).

# Why another modal plugin?

Most plugins I've found try to do too much, and have specialized ways of handling photo galleries, iframes and video.  The resulting HTML & CSS is often bloated and difficult to customize.

By contrast, this plugin handles the two most common scenarios I run into

* displaying an existing DOM element
* loading a page with AJAX

and does so with as little HTML & CSS as possible.

# Installation

You can install [jquery-modal](https://www.npmjs.com/package/jquery-modal) with npm:

`npm install jquery-modal`

or with [Bower](http://bower.io/):

`bower install jquery-modal`

or use the hosted version from [cdnjs](https://cdnjs.com/libraries/jquery-modal):

```html
<!-- Remember to include jQuery :) -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.0.0/jquery.min.js"></script>

<!-- jQuery Modal -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-modal/0.9.1/jquery.modal.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery-modal/0.9.1/jquery.modal.min.css" />
```

**Using Rails?** Check out the [jquery-modal-rails plugin](https://github.com/dei79/jquery-modal-rails)!

**jQuery Requirements:** As of version 0.3.0, jQuery 1.7 is required. If you're using an earlier version of jQuery you can use the [v.0.2.5 tag.](https://github.com/kylefox/jquery-modal/tags)

**Naming conflict with Bootstrap:** Bootstrap's [modal](http://getbootstrap.com/javascript/#modals) uses the same `$.modal` namespace. If you want to use jquery-modal with Bootstrap, the simplest solution is to manually modify the name of this plugin.

# Opening

#### Method 1: Automatically attaching to links

The simplest approach is to add `rel="modal:open"` to your links and use the `href` attribute to specify what to open in the modal.

Open an existing DOM element by ID:

```html
<form id="login-form" class="modal">
  ...
</form>

<a href="#login-form" rel="modal:open">Login</a>
```

Load a remote URL with AJAX:

```html
<a href="login.html" rel="modal:open">Login</a>
```

#### Method 2: Manually

You can manually open a modal by calling the `.modal()` method on the element:

```html
<form id="login-form" class="modal">
  ...
</form>
```

```js
$('#login-form').modal();
```

You can also invoke `.modal()` directly on links:

```html
<a href="#ex5" data-modal>Open a DOM element</a>
<a href="ajax.html" data-modal>Open an AJAX modal</a>
```

```js
$('a[data-modal]').click(function(event) {
  $(this).modal();
  return false;
});
```

### Compatibility Fallback

You can provide a clean fallback for users who have JavaScript disabled by manually attaching the modal via the `data-modal` attribute. This allows you to write your links pointing to the `href` as normal (fallback) while enabling modals where JavaScript is enabled.

```html
<!-- By default link takes user to /login.html -->
<a href="/login.html" data-modal="#login-modal">Login</a>

<!-- Login modal embedded in page -->
<div id="login-modal" class="modal">
  ...
</div>

<!-- For browsers with JavaScript, open the modal. -->
<script>
  $(function() {
    $('a[data-modal]').on('click', function() {
      $($(this).data('modal')).modal();
      return false;
    });
  });
</script>
```

#### Fade Transitions

By default the overlay & window appear instantaneously, but you can enable a fade effect by specifying the `fadeDuration` option.

```js
$('a.open-modal').click(function(event) {
  $(this).modal({
    fadeDuration: 250
  });
  return false;
});
```

This will fade in the overlay and modal over 250 milliseconds _simultaneously._ If you want the effect of the overlay appearing _before_ the window, you can specify the `fadeDelay` option. This indicates at what point during the overlay transition the window transition should begin.

So if you wanted the window to fade in when the overlay's was 80% finished:

```js
$(elm).modal({
  fadeDuration: 250,
  fadeDelay: 0.80
});
```

Or, if you wanted the window to fade in a few moments after the overlay transition has completely finished:

```js
$(elm).modal({
  fadeDuration: 250,
  fadeDelay: 1.5
});
```

The `fadeDelay` option only applies when opening the modal. When closing the modal, both the modal and the overlay fade out simultaneously according to the `fadeDuration` setting.

Fading is the only supported transition.

# Closing

Because there can be only one modal active at a single time, there's no need to select which modal to close:

```js
$.modal.close();
```

Similar to how links can be automatically bound to open modals, they can be bound to close modals using `rel="modal:close"`:

```html
<a href="#close" rel="modal:close">Close window</a>
```

_(Note that modals loaded with AJAX are removed from the DOM when closed)._

# Checking current state

* Use `$.modal.isActive()` to check if a modal is currently being displayed.
* Use `$.modal.getCurrent()` to retrieve a reference to the currently active modal instance, if any.

# Options

These are the supported options and their default values:

```js
$.modal.defaults = {
  closeExisting: true,    // Close existing modals. Set this to false if you need to stack multiple modal instances.
  escapeClose: true,      // Allows the user to close the modal by pressing `ESC`
  clickClose: true,       // Allows the user to close the modal by clicking the overlay
  closeText: 'Close',     // Text content for the close <a> tag.
  closeClass: '',         // Add additional class(es) to the close <a> tag.
  showClose: true,        // Shows a (X) icon/link in the top-right corner
  modalClass: "modal",    // CSS class added to the element being displayed in the modal.
  blockerClass: "modal",  // CSS class added to the overlay (blocker).

  // HTML appended to the default spinner during AJAX requests.
  spinnerHtml: '<div class="rect1"></div><div class="rect2"></div><div class="rect3"></div><div class="rect4"></div>',

  showSpinner: true,      // Enable/disable the default spinner during AJAX requests.
  fadeDuration: null,     // Number of milliseconds the fade transition takes (null means no transition)
  fadeDelay: 1.0          // Point during the overlay's fade-in that the modal begins to fade in (.5 = 50%, 1.5 = 150%, etc.)
};
```

# Events

The following events are triggered on the modal element at various points in the open/close cycle (see below for AJAX events).

```javascript
$.modal.BEFORE_BLOCK = 'modal:before-block';    // Fires just before the overlay (blocker) appears.
$.modal.BLOCK = 'modal:block';                  // Fires after the overlay (block) is visible.
$.modal.BEFORE_OPEN = 'modal:before-open';      // Fires just before the modal opens.
$.modal.OPEN = 'modal:open';                    // Fires after the modal has finished opening.
$.modal.BEFORE_CLOSE = 'modal:before-close';    // Fires when the modal has been requested to close.
$.modal.CLOSE = 'modal:close';                  // Fires when the modal begins closing (including animations).
$.modal.AFTER_CLOSE = 'modal:after-close';      // Fires after the modal has fully closed (including animations).
```

The first and only argument passed to these event handlers is the `modal` object, which has four properties:

```js
modal.$elm;       // Original jQuery object upon which modal() was invoked.
modal.options;    // Options passed to the modal.
modal.$blocker;   // The overlay element.
modal.$anchor;    // Anchor element originating the event.
```

So, you could do something like this:

```js
$('#purchase-form').on($.modal.BEFORE_CLOSE, function(event, modal) {
  clear_shopping_cart();
});
```

# AJAX

## Basic support

jQuery Modal uses $.get for basic AJAX support. A simple spinner will be displayed by default (if you've included modal.css) and will have the class `modal-spinner`. If you've set the `modalClass` option, the spinner will be prefixed with that class name instead.

You can add text or additional HTML to the spinner with the `spinnerHtml` option, or disable the spinner entirely by setting `showSpinner: false`.

The default spinner is from the excellent [SpinKit](http://tobiasahlin.com/spinkit/) by [Tobias Ahlin](https://twitter.com/tobiasahlin) 👍

## Events

The following events are triggered when AJAX modals are requested.

```js
$.modal.AJAX_SEND = 'modal:ajax:send';
$.modal.AJAX_SUCCESS = 'modal:ajax:success';
$.modal.AJAX_FAIL = 'modal:ajax:fail';
$.modal.AJAX_COMPLETE = 'modal:ajax:complete';
```

The handlers receive no arguments. The events are triggered on the `<a>` element which initiated the AJAX modal.

## More advanced AJAX handling

It's a good idea to provide more robust AJAX handling -- error handling, in particular. Instead of accommodating the myriad [`$.ajax` options](http://api.jquery.com/jQuery.ajax/) jQuery provides, jquery-modal makes it possible to directly modify the AJAX request itself.

Simply bypass the default AJAX handling (i.e.: don't use `rel="modal"`)

```html
<a href="ajax.html" rel="ajax:modal">Click me!</a>
```

and make your AJAX request in the link's click handler. Note that you need to manually append the new HTML/modal in the `success` callback:

```js
$('a[rel="ajax:modal"]').click(function(event) {

  $.ajax({

    url: $(this).attr('href'),

    success: function(newHTML, textStatus, jqXHR) {
      $(newHTML).appendTo('body').modal();
    },

    error: function(jqXHR, textStatus, errorThrown) {
      // Handle AJAX errors
    }

    // More AJAX customization goes here.

  });

  return false;
});
```

Note that the AJAX response must be wrapped in a div with class <code>modal</code> when using the second (manual) method.

# Bugs & Feature Requests

### Found a bug? MEH!

![](http://drops.kylefox.ca/1cqGP+)

**Just kidding.** Please [create an issue](https://github.com/kylefox/jquery-modal/issues/new) and **include a publicly-accessible demonstration of the bug.** [Dropbox](https://www.dropbox.com) or [JSFiddle](http://jsfiddle.net/) work well for demonstrating reproducable bugs, but you can use anything as long as it's publicly accessible. Your issue is much more likely to be resolved/merged if it includes a fix & pull request.

**Have an idea that improves jquery-modal?** Awesome! Please fork this repository, implement your idea (including documentation, if necessary), and submit a pull request.

I don't use this library as frequently as I used to, so if you want to see a fix/improvement you're best off submitting a pull request. Bugs without a test case and/or improvements without a pull request will be shown no mercy and closed!

# Contributing

## Maintainers Wanted

![](https://img.shields.io/badge/maintainers-wanted-red.svg)

This library became more popular and active than I ever expected, and unfortunately I don't have time to maintain it myself.

If you are interested in helping me maintain this library, please let me know — **I would love your help!**

[**Read more about becoming a maintainer &raquo;**](https://github.com/kylefox/jquery-modal/issues/170)

_I'd especially like people who would be excited about working towards a brand new **jQuery Modal 2.0**. See my [Proposal for jQuery Modal 2.0](https://github.com/kylefox/jquery-modal/issues/169) for more details & discussion._

## How to contribute

I welcome improvements to this plugin, particularly with:

* Performance improvements
* Making the code as concise/efficient as possible
* Bug fixes & browser compatibility

Please fork and send pull requests, or create an [issue](https://github.com/kylefox/jquery-modal/issues). Keep in mind the spirit of this plugin is **minimalism** so I'm very picky about adding _new_ features.

## Tips for development/contributing

* Make sure dependencies are installed: `npm install`
* After modifying `jquery.modal.js` and/or `jquery.modal.css`, you can optionally regenerate the minified files with `gulp min` and `gulp css` respectively.
* Make sure you have updated documentation (`README.md` and/or `examples/index.html`) if necessary. **Pull requests without documentation updates will be rejected.**
* Maintainers should increment version numbers and run `gulp changelog` when cutting a new release.

# Support

Please post a question on [StackOverflow](http://stackoverflow.com/). Commercial support by email is also available — please contact kylefox@gmail.com for rates. Unfortunately I am unable to provide free email support.

# License (MIT)

jQuery Modal is distributed under the [MIT License](Learn more at http://opensource.org/licenses/mit-license.php):

    Copyright (c) 2012 Kyle Fox

    Permission is hereby granted, free of charge, to any person obtaining
    a copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
    LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
    OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
    WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
<a name="0.9.2"></a>
## 0.9.2 (2019-02-25)


* Added anchor to custom event context ([3770312](https://github.com/kylefox/jquery-modal/commit/3770312))
* Merge pull request #238 from artem-tim/master ([30875a5](https://github.com/kylefox/jquery-modal/commit/30875a5))
* Updated documentation & min js ([89528ab](https://github.com/kylefox/jquery-modal/commit/89528ab))



<a name="0.8.2"></a>
## 0.8.2 (2017-07-03)


* add anchors to examples ([bb623e3](https://github.com/kylefox/jquery-modal/commit/bb623e3))
* Added syntax modes to code samples (closes #47) ([97c043a](https://github.com/kylefox/jquery-modal/commit/97c043a)), closes [#47](https://github.com/kylefox/jquery-modal/issues/47)
* alias elm as $elm (fixes #205) ([c074130](https://github.com/kylefox/jquery-modal/commit/c074130)), closes [#205](https://github.com/kylefox/jquery-modal/issues/205)
* Bind to links with multiple `rel` values ([c6fc99c](https://github.com/kylefox/jquery-modal/commit/c6fc99c)), closes [#135](https://github.com/kylefox/jquery-modal/issues/135)
* Bump to 0.8.1 ([35b84c4](https://github.com/kylefox/jquery-modal/commit/35b84c4))
* Bump to 0.8.2 ([196c43d](https://github.com/kylefox/jquery-modal/commit/196c43d))
* Changelog 0.8.1 ([7a10b52](https://github.com/kylefox/jquery-modal/commit/7a10b52))
* fix merge conflict after 0.8.1 release ([5b83154](https://github.com/kylefox/jquery-modal/commit/5b83154))
* Merge branch 'master' of github.com:kylefox/jquery-modal ([beb02fa](https://github.com/kylefox/jquery-modal/commit/beb02fa))
* Merge branch 'sbussetti-master' ([2428d64](https://github.com/kylefox/jquery-modal/commit/2428d64))
* Update compositor.json ([511dd72](https://github.com/kylefox/jquery-modal/commit/511dd72))
* Update compositor.json ([6f9ebcb](https://github.com/kylefox/jquery-modal/commit/6f9ebcb))
* Update README.md ([334b18a](https://github.com/kylefox/jquery-modal/commit/334b18a))



<a name="0.8.1"></a>
## 0.8.1 (2017-07-03)


* add anchors to examples ([bb623e3](https://github.com/kylefox/jquery-modal/commit/bb623e3))
* Added syntax modes to code samples (closes #47) ([97c043a](https://github.com/kylefox/jquery-modal/commit/97c043a)), closes [#47](https://github.com/kylefox/jquery-modal/issues/47)
* alias elm as $elm (fixes #205) ([c074130](https://github.com/kylefox/jquery-modal/commit/c074130)), closes [#205](https://github.com/kylefox/jquery-modal/issues/205)
* Bind to links with multiple `rel` values ([c6fc99c](https://github.com/kylefox/jquery-modal/commit/c6fc99c)), closes [#135](https://github.com/kylefox/jquery-modal/issues/135)
* Bump to 0.8.1 ([35b84c4](https://github.com/kylefox/jquery-modal/commit/35b84c4))
* Merge branch 'master' of github.com:kylefox/jquery-modal ([beb02fa](https://github.com/kylefox/jquery-modal/commit/beb02fa))
* Update compositor.json ([511dd72](https://github.com/kylefox/jquery-modal/commit/511dd72))
* Update compositor.json ([6f9ebcb](https://github.com/kylefox/jquery-modal/commit/6f9ebcb))
* Update README.md ([334b18a](https://github.com/kylefox/jquery-modal/commit/334b18a))



<a name="0.8.0"></a>
# 0.8.0 (2016-09-11)


* Bump to 0.8.0 ([1bc2a6b](https://github.com/kylefox/jquery-modal/commit/1bc2a6b))
* document $.modal.getCurrent() ([1310ef1](https://github.com/kylefox/jquery-modal/commit/1310ef1))
* Make modal more responsive-friendly ([b721dda](https://github.com/kylefox/jquery-modal/commit/b721dda))



<a name="0.7.3"></a>
## 0.7.3 (2016-09-11)


* Add $.modal.getCurrent() to retrieve current instance (if any) ([aee8564](https://github.com/kylefox/jquery-modal/commit/aee8564)), closes [#178](https://github.com/kylefox/jquery-modal/issues/178)



<a name="0.7.2"></a>
## 0.7.2 (2016-09-11)


* Bump to 0.7.2 ([51575f3](https://github.com/kylefox/jquery-modal/commit/51575f3))
* contributing notes (resolves #164) ([b07d3ba](https://github.com/kylefox/jquery-modal/commit/b07d3ba)), closes [#164](https://github.com/kylefox/jquery-modal/issues/164)
* log AFTER_CLOSE event in examples ([019f03e](https://github.com/kylefox/jquery-modal/commit/019f03e))
* Merge pull request #177 from shane-tomlinson/jquery-3.1 ([b49206b](https://github.com/kylefox/jquery-modal/commit/b49206b))
* Support for jQuery 3.x ([0576775](https://github.com/kylefox/jquery-modal/commit/0576775))
* Update README.md ([9448759](https://github.com/kylefox/jquery-modal/commit/9448759))
* Update README.md ([9e5491b](https://github.com/kylefox/jquery-modal/commit/9e5491b))
* Update README.md ([e345ddb](https://github.com/kylefox/jquery-modal/commit/e345ddb))
* Update README.md ([6e50636](https://github.com/kylefox/jquery-modal/commit/6e50636))
* Update to jQuery 3.1.0 in examples ([181b31f](https://github.com/kylefox/jquery-modal/commit/181b31f))



<a name="0.7.1"></a>
## 0.7.1 (2016-08-18)


* add minified css ([4f43415](https://github.com/kylefox/jquery-modal/commit/4f43415))
* add snippet to be able to load jquery as a module ([a3f1508](https://github.com/kylefox/jquery-modal/commit/a3f1508))
* add the minified javascript source ([27c88ac](https://github.com/kylefox/jquery-modal/commit/27c88ac))
* Explain events. ([99034ae](https://github.com/kylefox/jquery-modal/commit/99034ae))
* Merge pull request #154 from YuriBrunetto/min-css ([8a45590](https://github.com/kylefox/jquery-modal/commit/8a45590))
* Merge pull request #166 from Cryde/master ([85267db](https://github.com/kylefox/jquery-modal/commit/85267db))



<a name="0.7.0"></a>
# 0.7.0 (2016-03-17)


### NEW

* Added a demo for closeExisting ([b311bb5](https://github.com/kylefox/jquery-modal/commit/b311bb5))
* CSS tweak: removed opacity on .behind > .modal ([35db0c4](https://github.com/kylefox/jquery-modal/commit/35db0c4))
* Updated README.md ([c87c3e4](https://github.com/kylefox/jquery-modal/commit/c87c3e4))
* NEW: closeExisting property ([c6c767e](https://github.com/kylefox/jquery-modal/commit/c6c767e))

* Update index.html ([f5dbf0e](https://github.com/kylefox/jquery-modal/commit/f5dbf0e))

jquery.modal.pack.js
/*
    A simple jQuery modal (http://github.com/kylefox/jquery-modal)
    Version 0.2.2
*/
eval(function(p,a,c,k,e,d){e=function(c){return(c<a?'':e(parseInt(c/a)))+((c=c%a)>35?String.fromCharCode(c+29):c.toString(36))};if(!''.replace(/^/,String)){while(c--){d[e(c)]=k[c]||e(c)}k=[function(e){return d[e]}];e=function(){return'\\w+'};c=1};while(c--){if(k[c]){p=p.replace(new RegExp('\\b'+e(c)+'\\b','g'),k[c])}}return p}('(9(){n 4=x;$.5.3=9(7){n $6=$(A);b($6.W(\'m\')){$6.j(r);B}7=$.15({},$.5.3.P,7);9 g(){4.e=$(\'<Q U="14-3 e"></Q>\').Y({L:0,13:0,16:0,K:0,17:"Z%",I:"Z%",S:"R",f:7.f,1a:7.N,s:7.s});b(7.10){$(19).12(9(c){b(c.18==1b){$.5.3.8()}})}b(7.F){4.e.j($.5.3.8)}$(\'H\').M(4.e);$6.d($.5.3.D,[4])}9 u(){k(4);b(7.v){4.i=$(\'<a m="#8-3" t="3:8" U="8-3">11</a>\');4.6.M(4.i)}$6.V(7.X).V(\'1i\').u();$6.d($.5.3.z,[4])}4={6:$6,7:7};$6.d($.5.3.y,[4]);g();$6.d($.5.3.C,[4]);u()};$.5.3.P={N:"#1k",s:0.1m,f:1,10:q,F:q,X:"3",v:q};$.5.3.y=\'3:o-g\';$.5.3.D=\'3:g\';$.5.3.C=\'3:o-p\';$.5.3.z=\'3:p\';$.5.3.E=\'3:o-8\';$.5.3.G=\'3:8\';$.5.3.8=9(c){b(c){c.w()}b(!4){B}4.6.d($.5.3.E,[4]);b(4.i){4.i.l()}4.e.l();4.6.1o();4.6.d($.5.3.G,[4]);4=x};$.5.3.1c=9(){k(4)};9 r(c){c.w();n h=$(A).W(\'m\');b(h.1n(/^#/)){$(h).3()}1l{$.1j(h,{},9(T){$(T).1e(\'H\').1d(\'3:8\',9(c,3){3.6.l()}).3()})}}9 k(3){3.6.Y({S:\'R\',L:"J%",K:"J%",1f:-(3.6.I()/2),1g:-(3.6.1h()/2),f:3.7.f+1})};$(\'a[t="3:p"]\').O(\'j\',r);$(\'a[t="3:8"]\').O(\'j\',$.5.3.8)})();',62,87,'|||modal|current_modal|fn|elm|options|close|function||if|event|trigger|blocker|zIndex|block|target|closeButton|click|center_modal|remove|href|var|before|open|true|open_modal_from_link|opacity|rel|show|showClose|preventDefault|null|BEFORE_BLOCK|OPEN|this|return|BEFORE_OPEN|BLOCK|BEFORE_CLOSE|clickClose|CLOSE|body|height|50|left|top|append|overlay|live|defaults|div|fixed|position|html|class|addClass|attr|modalClass|css|100|escapeClose|Close|keydown|right|jquery|extend|bottom|width|which|document|background|27|resize|bind|appendTo|marginTop|marginLeft|outerWidth|current|get|000|else|75|match|hide'.split('|'),0,{}))


jquery.modal.js
/*
    A simple jQuery modal (http://github.com/kylefox/jquery-modal)
    Version 0.9.2
*/

(function (factory) {
  // Making your jQuery plugin work better with npm tools
  // http://blog.npmjs.org/post/112712169830/making-your-jquery-plugin-work-better-with-npm
  if(typeof module === "object" && typeof module.exports === "object") {
    factory(require("jquery"), window, document);
  }
  else {
    factory(jQuery, window, document);
  }
}(function($, window, document, undefined) {

  var modals = [],
      getCurrent = function() {
        return modals.length ? modals[modals.length - 1] : null;
      },
      selectCurrent = function() {
        var i,
            selected = false;
        for (i=modals.length-1; i>=0; i--) {
          if (modals[i].$blocker) {
            modals[i].$blocker.toggleClass('current',!selected).toggleClass('behind',selected);
            selected = true;
          }
        }
      };

  $.modal = function(el, options) {
    var remove, target;
    this.$body = $('body');
    this.options = $.extend({}, $.modal.defaults, options);
    this.options.doFade = !isNaN(parseInt(this.options.fadeDuration, 10));
    this.$blocker = null;
    if (this.options.closeExisting)
      while ($.modal.isActive())
        $.modal.close(); // Close any open modals.
    modals.push(this);
    if (el.is('a')) {
      target = el.attr('href');
      this.anchor = el;
      //Select element by id from href
      if (/^#/.test(target)) {
        this.$elm = $(target);
        if (this.$elm.length !== 1) return null;
        this.$body.append(this.$elm);
        this.open();
      //AJAX
      } else {
        this.$elm = $('<div>');
        this.$body.append(this.$elm);
        remove = function(event, modal) { modal.elm.remove(); };
        this.showSpinner();
        el.trigger($.modal.AJAX_SEND);
        $.get(target).done(function(html) {
          if (!$.modal.isActive()) return;
          el.trigger($.modal.AJAX_SUCCESS);
          var current = getCurrent();
          current.$elm.empty().append(html).on($.modal.CLOSE, remove);
          current.hideSpinner();
          current.open();
          el.trigger($.modal.AJAX_COMPLETE);
        }).fail(function() {
          el.trigger($.modal.AJAX_FAIL);
          var current = getCurrent();
          current.hideSpinner();
          modals.pop(); // remove expected modal from the list
          el.trigger($.modal.AJAX_COMPLETE);
        });
      }
    } else {
      this.$elm = el;
      this.anchor = el;
      this.$body.append(this.$elm);
      this.open();
    }
  };

  $.modal.prototype = {
    constructor: $.modal,

    open: function() {
      var m = this;
      this.block();
      this.anchor.blur();
      if(this.options.doFade) {
        setTimeout(function() {
          m.show();
        }, this.options.fadeDuration * this.options.fadeDelay);
      } else {
        this.show();
      }
      $(document).off('keydown.modal').on('keydown.modal', function(event) {
        var current = getCurrent();
        if (event.which === 27 && current.options.escapeClose) current.close();
      });
      if (this.options.clickClose)
        this.$blocker.click(function(e) {
          if (e.target === this)
            $.modal.close();
        });
    },

    close: function() {
      modals.pop();
      this.unblock();
      this.hide();
      if (!$.modal.isActive())
        $(document).off('keydown.modal');
    },

    block: function() {
      this.$elm.trigger($.modal.BEFORE_BLOCK, [this._ctx()]);
      this.$body.css('overflow','hidden');
      this.$blocker = $('<div class="' + this.options.blockerClass + ' blocker current"></div>').appendTo(this.$body);
      selectCurrent();
      if(this.options.doFade) {
        this.$blocker.css('opacity',0).animate({opacity: 1}, this.options.fadeDuration);
      }
      this.$elm.trigger($.modal.BLOCK, [this._ctx()]);
    },

    unblock: function(now) {
      if (!now && this.options.doFade)
        this.$blocker.fadeOut(this.options.fadeDuration, this.unblock.bind(this,true));
      else {
        this.$blocker.children().appendTo(this.$body);
        this.$blocker.remove();
        this.$blocker = null;
        selectCurrent();
        if (!$.modal.isActive())
          this.$body.css('overflow','');
      }
    },

    show: function() {
      this.$elm.trigger($.modal.BEFORE_OPEN, [this._ctx()]);
      if (this.options.showClose) {
        this.closeButton = $('<a href="#close-modal" rel="modal:close" class="close-modal ' + this.options.closeClass + '">' + this.options.closeText + '</a>');
        this.$elm.append(this.closeButton);
      }
      this.$elm.addClass(this.options.modalClass).appendTo(this.$blocker);
      if(this.options.doFade) {
        this.$elm.css({opacity: 0, display: 'inline-block'}).animate({opacity: 1}, this.options.fadeDuration);
      } else {
        this.$elm.css('display', 'inline-block');
      }
      this.$elm.trigger($.modal.OPEN, [this._ctx()]);
    },

    hide: function() {
      this.$elm.trigger($.modal.BEFORE_CLOSE, [this._ctx()]);
      if (this.closeButton) this.closeButton.remove();
      var _this = this;
      if(this.options.doFade) {
        this.$elm.fadeOut(this.options.fadeDuration, function () {
          _this.$elm.trigger($.modal.AFTER_CLOSE, [_this._ctx()]);
        });
      } else {
        this.$elm.hide(0, function () {
          _this.$elm.trigger($.modal.AFTER_CLOSE, [_this._ctx()]);
        });
      }
      this.$elm.trigger($.modal.CLOSE, [this._ctx()]);
    },

    showSpinner: function() {
      if (!this.options.showSpinner) return;
      this.spinner = this.spinner || $('<div class="' + this.options.modalClass + '-spinner"></div>')
        .append(this.options.spinnerHtml);
      this.$body.append(this.spinner);
      this.spinner.show();
    },

    hideSpinner: function() {
      if (this.spinner) this.spinner.remove();
    },

    //Return context for custom events
    _ctx: function() {
      return { elm: this.$elm, $elm: this.$elm, $blocker: this.$blocker, options: this.options, $anchor: this.anchor };
    }
  };

  $.modal.close = function(event) {
    if (!$.modal.isActive()) return;
    if (event) event.preventDefault();
    var current = getCurrent();
    current.close();
    return current.$elm;
  };

  // Returns if there currently is an active modal
  $.modal.isActive = function () {
    return modals.length > 0;
  };

  $.modal.getCurrent = getCurrent;

  $.modal.defaults = {
    closeExisting: true,
    escapeClose: true,
    clickClose: true,
    closeText: 'Close',
    closeClass: '',
    modalClass: "modal",
    blockerClass: "jquery-modal",
    spinnerHtml: '<div class="rect1"></div><div class="rect2"></div><div class="rect3"></div><div class="rect4"></div>',
    showSpinner: true,
    showClose: true,
    fadeDuration: null,   // Number of milliseconds the fade animation takes.
    fadeDelay: 1.0        // Point during the overlay's fade-in that the modal begins to fade in (.5 = 50%, 1.5 = 150%, etc.)
  };

  // Event constants
  $.modal.BEFORE_BLOCK = 'modal:before-block';
  $.modal.BLOCK = 'modal:block';
  $.modal.BEFORE_OPEN = 'modal:before-open';
  $.modal.OPEN = 'modal:open';
  $.modal.BEFORE_CLOSE = 'modal:before-close';
  $.modal.CLOSE = 'modal:close';
  $.modal.AFTER_CLOSE = 'modal:after-close';
  $.modal.AJAX_SEND = 'modal:ajax:send';
  $.modal.AJAX_SUCCESS = 'modal:ajax:success';
  $.modal.AJAX_FAIL = 'modal:ajax:fail';
  $.modal.AJAX_COMPLETE = 'modal:ajax:complete';

  $.fn.modal = function(options){
    if (this.length === 1) {
      new $.modal(this, options);
    }
    return this;
  };

  // Automatically bind links with rel="modal:close" to, well, close the modal.
  $(document).on('click.modal', 'a[rel~="modal:close"]', $.modal.close);
  $(document).on('click.modal', 'a[rel~="modal:open"]', function(event) {
    event.preventDefault();
    $(this).modal();
  });
}));
